File: frontend/src/types/workspace.ts
export type WorkspaceRole = 'owner' | 'admin' | 'editor' | 'commenter' | 'viewer'
export type Workspace = { id: string; name: string; description: string; owner_id: string; visibility: 'public' | 'private'; is_password_protected: boolean; role?: WorkspaceRole | null; member_count: number; created_at: string; updated_at: string }
export type WorkspaceMember = { user_id: string; name: string; email: string; role: WorkspaceRole; joined_at: string }

File: backend/app/models/workspace.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

WorkspaceRole = Literal["owner", "editor", "viewer"]
WorkspaceVisibility = Literal["public", "private"]


class Workspace(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str = Field(alias="_id")
    name: str
    description: str = ""
    owner_id: str
    visibility: WorkspaceVisibility
    password_hash: str | None = None
    created_at: datetime
    updated_at: datetime


class WorkspaceMember(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str = Field(alias="_id")
    workspace_id: str
    user_id: str
    role: WorkspaceRole
    joined_at: datetime
